### PR TITLE
Disable default linear tearing

### PR DESCRIPTION
**Pull request by AnHeuermann**
For [track ticket 6196](https://trac.openmodelica.org/OpenModelica/ticket/6196).

`maxSizeLinearTearing` now defaults to 0 which disables linear tearing.

### CODE REVIEW

**Comment by casella** [comment]
Looks good. I would wait one or two days before merging, until the situation with the Jenkins reporting is stabilized. Otherwise it may be difficult to ascertain what is the impact of this change on the larger testsuite.

**Comment by casella** [comment]
@AnHeuermann, @kabdelhak , one more comment. 

Linear tearing was supposedly used to reduce the complexity of linear system. I have an interesting example that stems from object-oriented, finite-volume discretization of the heat equation in solids. Suppose you have this linear system with two equations
```
    Q = 2*g*(Tb - T2);
    Q = 2*g*(T1 - Tb);
```
with unknowns `Q` and `Tb` (`T1` and `T2` are states)

The closed-form solution of this sytem is 
```
Q := g*(T1 - T2);
Tb := (T1 + T2)/2;
```
which is very convenient to compute, instead of resorting to LAPACK. In general it is not a good idea to use the analytical solution of a 2x2 linear system, compared to using a good linear solver with gaussian elimination and pivoting, but in this specific case the symbolic solution is just perfect.

Is there any way to get this in OpenModelica? Should I open a discussion ticket about that?

**Comment by AnHeuermann** [comment]
> Is there any way to get this in OpenModelica?

Sounds very hard. So my guess would be that any sort of optimization is very expensive.

Tearing is NP-complete (if I remember correctly). I'm not sure what complexity our heuristics are but I guess at least `O(n^2)`.
So that would be the complexity to beat to do this sort of improvement.
I could not find a exact statement about the LAPACK solver we are using just now, but I think it is around `O(n^2)` as well.

But if you have any idea please open a discussion ticket and we can play around with this for a bit.

**Comment by kabdelhak** [comment]
We could use the ASSC algorithm which tries to solve those systems with gaussian elimination. Currently it is only used to convert analytical to structural singularity, but with `-d=fullASSC` it will convert the linear-integer valued part of algebraic loops to triangular form. We could update the algorithm to also be able to handle real valued coefficents... Please note that taking parameters into account would probably need way more work.

If you have an example for which i could test this it would be great!

**Comment by AnHeuermann** [comment]
I should probably disable the warning if `maxSizeLinearTearing=0` if we disable linear tearing for now.

```bash
Notification: Tearing is skipped for strong component 5 because system size of 2 exceeds maximum system size for tearing of linear systems (0).
To adjust the maximum system size for tearing use --maxSizeLinearTearing=<size> and --maxSizeNonlinearTearing=<size>.
```

**Comment by casella** [comment]
> > Is there any way to get this in OpenModelica?
> 
> Sounds very hard. So my guess would be that any sort of optimization is very expensive.
> 
> Tearing is NP-complete (if I remember correctly). I'm not sure what complexity our heuristics are but I guess at least `O(n^2)`.
> So that would be the complexity to beat to do this sort of improvement.

Mind you, the idea is to limit this kind of optimization to 2x2, max 3x3 systems, after removing aliases, that is. The likelyhood that larger systems have a small, efficient closed-form solution is very low. So, there is no scaling issue here.

> I could not find a exact statement about the LAPACK solver we are using just now, but I think it is around `O(n^2)` as well.

Dense linear solvers have complexity [at most O(n^3) and above O(n^2)](https://simple.wikipedia.org/wiki/System_of_linear_equations).

**Comment by casella** [comment]
> We could use the ASSC algorithm which tries to solve those systems with gaussian elimination. Currently it is only used to convert analytical to structural singularity, but with `-d=fullASSC` it will convert the linear-integer valued part of algebraic loops to triangular form. We could update the algorithm to also be able to handle real valued coefficents...

Sounds good.

> Please note that taking parameters into account would probably need way more work.

We'll see. It would already good to have it without that, for extra optimization one should evaluate parameters anyway.

> If you have an example for which i could test this it would be great!

Well, the simplest example is shown above in my comment. As soon as we release our benchmarks, this equation will be replicated N*M*P times in a 3D thermal simulation.

**Comment by casella** [comment]
See [#6236](https://trac.openmodelica.org/OpenModelica/ticket/6236)

**Comment by casella** [comment]
> I should probably



> 
> 
> I should probably disable the warning if `maxSizeLinearTearing=0` if we disable linear tearing for now.
> 
> ```shell
> Notification: Tearing is skipped for strong component 5 because system size of 2 exceeds maximum system size for tearing of linear systems (0).
> To adjust the maximum system size for tearing use --maxSizeLinearTearing=<size> and --maxSizeNonlinearTearing=<size>.
> ```

Absolutely!

File: OMCompiler/Compiler/Util/Flags.mo [hunk 1]
```diff
@@ -1152,7 +1152,7 @@ constant ConfigFlag INST_CACHE_SIZE = CONFIG_FLAG(90, "instCacheSize",
   NONE(), EXTERNAL(), INT_FLAG(25343), NONE(),
   Gettext.gettext("Sets the size of the internal hash table used for instantiation caching."));
 constant ConfigFlag MAX_SIZE_LINEAR_TEARING = CONFIG_FLAG(91, "maxSizeLinearTearing",
-  NONE(), EXTERNAL(), INT_FLAG(200), NONE(),
+  NONE(), EXTERNAL(), INT_FLAG(0), NONE(),
   Gettext.gettext("Sets the maximum system size for tearing of linear systems (default 200)."));
```
Comment (phannebohm) [inline OMCompiler/Compiler/Util/Flags.mo:1156]:
```suggestion
  Gettext.gettext("Sets the maximum system size for tearing of linear systems (default 0, no tearing)."));
```

Comment (AnHeuermann) [reply OMCompiler/Compiler/Util/Flags.mo:1156]: Thanks for the catch. I totally forgot to change the description.

**Comment by phannebohm** [comment]
Wow, there are quite a lot of systems of size 2!.

**Comment by casella** [comment]
> Wow, there are quite a lot of systems of size 2!.

That's why I proposed [#6236](https://trac.openmodelica.org/OpenModelica/ticket/6236) 😃

**Comment by AnHeuermann** [comment]
We still have some problems left.
Setting `--maxSizeLinearTearing=0` prevents us from running minimal tearing on linear systems.
If minimal tearing would work perfect we should always use at least minimal tearing for every (linear) component.
But at the moment minimal tearing produces some teared loops even for examples where no tearing would be fine as well.

So I'll set minimal tearing to run when `--tearingMethod=minimalTearing` is set, even when `--maxSizeLinearTearing=0`. The same needs to be done for non-linear tearing. I'm doing this after #6976 is merged to have more clear changes.

**Comment by casella** [comment]
> We still have some problems left.
> Setting `--maxSizeLinearTearing=0` prevents us from running minimal tearing on linear systems.
> If minimal tearing would work perfect we should always use at least minimal tearing for every (linear) component.
> But at the moment minimal tearing produces some teared loops even for examples where no tearing would be fine as well.
> 
> So I'll set minimal tearing to run when `--tearingMethod=minimalTearing` is set, even when `--maxSizeLinearTearing=0`. The same needs to be done for non-linear tearing. I'm doing this after #6976 is merged to have more clear changes.

In that case, I'd suggest to also enable --tearingStrictness=veryStrict, to avoid bogus divisions by zero, see [#6232](https://trac.openmodelica.org/OpenModelica/ticket/6232)

**Comment by AnHeuermann** [comment]
@casella I'm not really sure this PR is still a good idea. We have some problems with tearing and chaining the default way at the moment could cause a lot of additional trouble.
I would need to change a bit more then just disabling the linear tearing but handle minimal tearing for the linear cases as well. And then everybody will end up using minimal tearing and I don'T think it is ready for that yet.

My suggestion: Let's try this again after 1.17

**Comment by casella** [comment]
> @casella I'm not really sure this PR is still a good idea. We have some problems with tearing and chaining the default way at the moment could cause a lot of additional trouble.
> I would need to change a bit more then just disabling the linear tearing but handle minimal tearing for the linear cases as well. And then everybody will end up using minimal tearing and I don'T think it is ready for that yet.
> 
> My suggestion: Let's try this again after 1.17

All deadlines are negotiable. However, this issue is very serious. We have many perfectly fine Modelica models in the testsuite which are failing because of this. Including models from PowerSysPro, which is why we started this discussion. This issue is not extremely urgent, but I won't postpone it to July 2021 or later. Let's try to find a good compromise.

One idea could be to make linear tearing veryStrict: simply don't flag an equation as solvable for a variable if that involves dividing by a parameter. It may still be the case that we are not picking the best pivot, so the solution is affected by higher rounding error than possible, but at least you avoid divisions by zero.

What do you think?

**Comment by AnHeuermann** [comment]
> One idea could be to make linear tearing veryStrict: simply don't flag an equation as solvable for a variable if that involves dividing by a parameter. It may still be the case that we are not picking the best pivot, so the solution is affected by higher rounding error than possible, but at least you avoid divisions by zero.

I started by activating minimal tearing for all linear loops when `minimalSizeLinearTearing=0`. With this I expect to solve most problems with discrete iteration variables but expect some bugs to pop up from minimal tearing (which is of course a good opportunity to solve them).
Having a more strict linear tearing could be a solution as well but ideally I would like to have minimal tearing be able to solve all linear loops in a satisfying fashion.

File: OMCompiler/Compiler/BackEnd/Tearing.mo [hunk 2]
```diff
@@ -337,7 +338,12 @@ algorithm
         maxSize := Flags.getConfigInt(Flags.MAX_SIZE_NONLINEAR_TEARING);
       end if;
 
-      useTearing := checkTearingSettings(maxSize, isLinear, strongComponentIndexOut, listLength(vindx));
+        // Use minimal tearing for linear components if linear tearing is disabled
+        if (Flags.getConfigInt(Flags.MAX_SIZE_LINEAR_TEARING) == 0 and isLinear) then
+          method := MINIMAL_TEARING();
+        end if;
```
Comment (phannebohm) [inline OMCompiler/Compiler/BackEnd/Tearing.mo:344]: One indent too much :slightly_smiling_face: 
```suggestion
      // Use minimal tearing for linear components if linear tearing is disabled
      if (Flags.getConfigInt(Flags.MAX_SIZE_LINEAR_TEARING) == 0 and isLinear) then
        method := MINIMAL_TEARING();
      end if;
```

**Review by phannebohm — APPROVED** [review]
I like it!

**Comment by casella** [comment]
> Having a more strict linear tearing could be a solution as well but ideally I would like to have minimal tearing be able to solve all linear loops in a satisfying fashion.

I have no strong opinions here. My strong requirement for 1.17.0 is that we no longer get division by zero when solving non-singular linear systems. As long as we get there, anything goes.

**Review by casella — APPROVED** [review]
@AnHeuermann, I am fine with this PR, I think it will solve a lot of problems with library testing. 

I see there is one error with the testsuite that prevents it to be merged. I couldn't figure out what it is.

Can you please try to solve it and merge this PR? I think it is important we have it in 1.17.0.

**Comment by AnHeuermann** [comment]
Okay, I'll try to finalize this tomorrow.

**Comment by casella** [comment]
> 
> 
> Okay, I'll try to finalize this tomorrow.

Wunderbar, vielen Dank!

**Comment by AnHeuermann** [comment]
Most failing tests are fixed now. The remaining failing tests are a bit more complicated.

For example `Modelica.Electrical.Machines.Examples.AsynchronousInductionMachines.AIMC_Transformer.mos – simulation_libraries_msl32` is still a problem. Without linear tearing the linear solver fails at every single time step and then the backup total pivot method is used. After around 6 minutes the test is aborted. I let it run during supper and it was still not finished after an hour. Before it was running in a few seconds.

I could cheat and add an `--maxSizeLinearTearing=200`, but then we will see the same results in the nightly builds.
When looking at the differences in the linear loops I see a lot of systems that increased in size (around 3 residual equations to 28 residual equations) and one system that was a `Equationsystem Jacobian Linear` before and is now a torn linear equation system.
The old system:
```
Equationsystem Jacobian Linear:
1: $DER.aimc.idq_sr[1]:VARIABLE(unit = "A" stateSelect=StateSelect.prefer final = true )  "Stator space phasor current / rotor fixed frame" type: Real [2]
2: $DER.aimc.idq_sr[2]:VARIABLE(unit = "A" stateSelect=StateSelect.prefer final = true )  "Stator space phasor current / rotor fixed frame" type: Real [2]
1/1 (1): $DER.aimc.lssigma.i_[2] = aimc.airGapS.RotationMatrix[2,1] * $DER.aimc.idq_sr[1] + $DER.aimc.airGapS.RotationMatrix[2,1] * aimc.idq_sr[1] + aimc.airGapS.RotationMatrix[2,2] * $DER.aimc.idq_sr[2] + $DER.aimc.airGapS.RotationMatrix[2,2] * aimc.idq_sr[2]   [dynamic |0|0|0|0|]
2/2 (1): $DER.aimc.lssigma.i_[1] = aimc.airGapS.RotationMatrix[2,2] * $DER.aimc.idq_sr[1] + $DER.aimc.airGapS.RotationMatrix[2,2] * aimc.idq_sr[1] + (-aimc.airGapS.RotationMatrix[2,1]) * $DER.aimc.idq_sr[2] - $DER.aimc.airGapS.RotationMatrix[2,1] * aimc.idq_sr[2]   [dynamic |0|0|0|0|]

Jac:
{1,1}:-aimc.airGapS.RotationMatrix[2,1],
{1,2}:-aimc.airGapS.RotationMatrix[2,2],
{2,1}:-aimc.airGapS.RotationMatrix[2,2],
{2,2}:aimc.airGapS.RotationMatrix[2,1]
```

The new system:
```

torn linear Equationsystem:

internal vars (0)

residual vars (2)
1: $DER.aimc.idq_sr[2]:VARIABLE(unit = "A" stateSelect=StateSelect.prefer final = true )  "Stator space phasor current / rotor fixed frame" type: Real [2]
2: $DER.aimc.idq_sr[1]:VARIABLE(unit = "A" stateSelect=StateSelect.prefer final = true )  "Stator space phasor current / rotor fixed frame" type: Real [2]

internal equations (0)

residual equations (2)
1/1 (1): $DER.aimc.lssigma.i_[1] = aimc.airGapS.RotationMatrix[2,2] * $DER.aimc.idq_sr[1] + $DER.aimc.airGapS.RotationMatrix[2,2] * aimc.idq_sr[1] + (-aimc.airGapS.RotationMatrix[2,1]) * $DER.aimc.idq_sr[2] - $DER.aimc.airGapS.RotationMatrix[2,1] * aimc.idq_sr[2]   [dynamic |0|0|0|0|]
2/2 (1): $DER.aimc.lssigma.i_[2] = aimc.airGapS.RotationMatrix[2,1] * $DER.aimc.idq_sr[1] + $DER.aimc.airGapS.RotationMatrix[2,1] * aimc.idq_sr[1] + aimc.airGapS.RotationMatrix[2,2] * $DER.aimc.idq_sr[2] + $DER.aimc.airGapS.RotationMatrix[2,2] * aimc.idq_sr[2]   [dynamic |0|0|0|0|]
```
@kabdelhak What is an `Equationsystem Jacobian Linear`?

I guess it is a left-over bug in minimal tearing.
Deactivating tearing (`--tearingMethod=noTearing`) for this model works wonders.

**Comment by AnHeuermann** [comment]
I'm trying to find the bug in minimal Tearing in [trac ticket #6285](https://trac.openmodelica.org/OpenModelica/ticket/6285#ticket).

**Comment by kabdelhak** [comment]
> @kabdelhak What is an `Equationsystem Jacobian Linear`?

We have two different types of strong components that represent algebraic loops. `BackendDAE.EQUATIONSYSTEM()` and `BackendDAE.TORNSYSTEM()`. The former one kind of is a legacy type before tearing was implemented. All systems of the first type after sorting and we compute a full jacobian (also a legacy type, now we have seed based generic jacobians) for those. If any of the loop variables are found in the jacobian it is considered nonlinear, else linear. Then we apply tearing and convert it to the latter type. Sometimes when tearing fails or any form of `noTearing` is applied, it will stick to the legacy type. That kinda works because we still have legacy code generation for those... but not maintained in any way since we introduced the latter type. 

What you got there probably is a BackendDAE.EQUATIONSYSTEM() with linear jacobian. Tearing failed or any form of `noTearing` was applied and it was not converted to the other type, which you definitely should do. You can do so even if you do not tear it at all, just copy all the equations and variables and create empty ones for inner equations and variables.

Maybe we should think about removing the code generation for legacy types and fix all the errors that come from it because they are only passing by pure chance.

**Comment by casella** [comment]
> For example `Modelica.Electrical.Machines.Examples.AsynchronousInductionMachines.AIMC_Transformer.mos – simulation_libraries_msl32` is still a problem. Without linear tearing the linear solver fails at every single time step and then the backup total pivot method is used. After around 6 minutes the test is aborted. I let it run during supper and it was still not finished after an hour. Before it was running in a few seconds.

This is obviously because of some bugs in the way the regular (i.e. non-torn) systems are solved. Of course solving a system with 28 equations will be slower than solving one with 3, _particularly if a sparse solver is not used_ (BTW, if that is the case, we should definitely re-tune the defaults for `-lssMaxDensity` and `-lssMinSize`, currently at 0.2 and 4001), but _it should not cause failures_.

@kabdelhak's comment seems to confirm this. I hope you can manage this without too much effort.

**Comment by AnHeuermann** [comment]
I'll rebase the master into this after https://github.com/OpenModelica/OpenModelica/pull/7122 is merged and see if that will solve most of the failing tests.

**Comment by casella** [comment]
I'm not sure we should push this into 1.17.0, too risky at this point in time, we need some time to test it thoroughly. What do you think?

**Comment by AnHeuermann** [comment]
> I'm not sure we should push this into 1.17.0, too risky at this point in time, we need some time to test it thoroughly. What do you think?

I agree. It will change basically every simulation that involves linear equation systems. There is no way this work right from the beginning.

**Comment by AnHeuermann** [comment]
This PR is stuck. There is no point in keeping it open.
I need to find some time and start from scratch.

**Pull request closed without merging**